hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java

```diff
@@ -2209,14 +2205,18 @@ private void listStatusFindKeyInTableCache(
           bucketName);
     }
 
+    List<OmKeyInfo> keyInfoList = new ArrayList<>(fileStatusList.size());
     for (OzoneFileStatus fileStatus : fileStatusList) {
-      if (args.getRefreshPipeline()) {
-        refreshPipeline(fileStatus.getKeyInfo());
-      }
-      if (args.getSortDatanodes()) {
+      keyInfoList.add(fileStatus.getKeyInfo());
+    }
+    refreshPipeline(keyInfoList);
+
+    if (args.getSortDatanodes()) {
+      for (OzoneFileStatus fileStatus : fileStatusList) {
         sortDatanodeInPipeline(fileStatus.getKeyInfo(), clientAddress);
```

Comment:
This call still makes number of entries in listStatus call to put nodes in sorted order in pipeline when ozone.network.topology.aware.read is set. (Default false, so this will not be a problem, but we have still the problem of making number of entries in listStatus calls to SCM.

Reply:
Thanks for highlighting this further possible improvement.  I'll file a separate Jira issue for this, as I think it's out of scope here.

Reply:
https://issues.apache.org/jira/browse/HDDS-4473

Reply:
Fine with me.